File: apps/entrelares_app/lib/models/care_schedule.dart
/// The `care_schedules` row. Mirrors `Entrelares/Models/CareSchedule.cs` —
/// including the T-33/T-35 concurrency contract: an authenticated UPDATE must
/// send the FULL row, with `revision` as read and `submitted_token` echoing
/// the `revision_token` that was read. Omitting the echo is what a forged
/// payload would do, and the trigger rejects it.
class CareSchedule {
  final int id;
  final DateTime scheduleDate; // date-only; ISO yyyy-MM-dd on the wire
  final String? handoffTime; // "HH:mm:ss" as PostgREST returns it
  final int scheduledParentId;
  final int? actualParentId;
  final String? notes;
  final String? createdAt;
  final String? updatedAt;
  final int revision;
  final String revisionToken;

  const CareSchedule({
    required this.id,
    required this.scheduleDate,
    this.handoffTime,
    required this.scheduledParentId,
    this.actualParentId,
    this.notes,
    this.createdAt,
    this.updatedAt,
    this.revision = 0,
    this.revisionToken = '',
  });

  int get effectiveParentId => actualParentId ?? scheduledParentId;

  static DateTime _parseDate(String iso) {
    final p = iso.split('-');
    return DateTime(int.parse(p[0]), int.parse(p[1]), int.parse(p[2]));
  }

  static String isoDate(DateTime d) =>
      '${d.year.toString().padLeft(4, '0')}-${d.month.toString().padLeft(2, '0')}-${d.day.toString().padLeft(2, '0')}';

  factory CareSchedule.fromJson(Map<String, dynamic> json) => CareSchedule(
        id: json['id'] as int,
        scheduleDate: _parseDate(json['schedule_date'] as String),
        handoffTime: json['handoff_time'] as String?,
        scheduledParentId: json['scheduled_parent_id'] as int,
        actualParentId: json['actual_parent_id'] as int?,
        notes: json['notes'] as String?,
        createdAt: json['created_at'] as String?,
        updatedAt: json['updated_at'] as String?,
        revision: (json['revision'] as int?) ?? 0,
        revisionToken: (json['revision_token'] as String?) ?? '',
      );

  /// INSERT payload: no id (identity), no tokens (server-stamped), no
  /// family_id (trigger_a derives it from scheduled_parent_id).
  Map<String, dynamic> toInsertJson() => {
        'schedule_date': isoDate(scheduleDate),
        'handoff_time': handoffTime,
        'scheduled_parent_id': scheduledParentId,
        'actual_parent_id': actualParentId,
        'notes': notes,
      };

  /// UPDATE payload: the full row, `revision` as read and `submitted_token`
  /// echoing the `revision_token` read — the only way past the T-35 guard.
  Map<String, dynamic> toUpdateJson() => {
        'schedule_date': isoDate(scheduleDate),
        'handoff_time': handoffTime,
        'scheduled_parent_id': scheduledParentId,
        'actual_parent_id': actualParentId,
        'notes': notes,
        'revision': revision,
        'submitted_token': revisionToken.isEmpty ? null : revisionToken,
      };

  CareSchedule copyWith({int? scheduledParentId, int? actualParentId}) =>
      CareSchedule(
        id: id,
        scheduleDate: scheduleDate,
        handoffTime: handoffTime,
        scheduledParentId: scheduledParentId ?? this.scheduledParentId,
        actualParentId: actualParentId ?? this.actualParentId,
        notes: notes,
        createdAt: createdAt,
        updatedAt: updatedAt,
        revision: revision,
        revisionToken: revisionToken,
      );
}

File: db-gate/Entrelares.DbContracts/Models/CareSchedule.cs
using Supabase.Postgrest.Attributes;
using Supabase.Postgrest.Models;

namespace Entrelares.Models
{
    [Table("care_schedules")]
    public class CareSchedule : BaseModel
    {
        [PrimaryKey("id", false)]
        public long Id { get; set; }

        [Column("schedule_date")]
        public DateOnly ScheduleDate { get; set; }

        [Column("handoff_time")]
        public TimeOnly? HandoffTime { get; set; }

        [Column("scheduled_parent_id")]
        public long ScheduledParentId { get; set; }

        [Column("actual_parent_id")]
        public long? ActualParentId { get; set; }

        [Column("notes")]
        public string? Notes { get; set; }

        [Column("created_at")]
        public DateTime CreatedAt { get; set; }

        [Column("updated_at")]
        public DateTime UpdatedAt { get; set; }

        // T-33: optimistic-concurrency counter. The DB increments it on every
        // UPDATE; a client update carrying a stale value is rejected with the
        // "someone got there first" message (trigger_c_enforce_schedule_revision).
        [Column("revision")]
        public int Revision { get; set; }

        // T-35: the non-guessable half of the same guard. The DB re-rolls this
        // token on every write, so it cannot be predicted from the row a client
        // already holds — unlike Revision, which a forged payload could simply
        // stamp as read+1.
        [Column("revision_token")]
        public Guid RevisionToken { get; set; }

        // T-35: the echo the DB checks against the row's current RevisionToken.
        // Computed, never stored: the trigger resets the column to NULL after
        // every write, and deriving it here (instead of assigning it at each
        // call site) is what makes it impossible for an update path to forget it
        // — every full-row Update() carries the token that was read, which is
        // the whole point of the guard. Omitting it is what a forged payload
        // would do, and the trigger rejects that.
        [Column("submitted_token")]
        public Guid? SubmittedToken
        {
            get => RevisionToken == Guid.Empty ? null : RevisionToken;
            set => _ = value;   // server-side value is always NULL; nothing to keep
        }
    }
}
